### ISSUE TITLE
[FEATURE] v3 single table attribute mapping

### ISSUE THREAD

**Issue by alexedwardjones** [dynamoose/dynamoose#1321]
### Summary:

I'm using the alpha of v3 to build out a single table design using multiple models each with their own schema. As a result of this single table design I'm using generic attributes (`pk`, `sk`, etc.) to store different things depending on the model. The problem is that you lose the meaning of these values on the different models.

Is there any plan to introduce a mapping of model attributes to underlying table attributes?

#### Example
```
const userSchema = new dynamoose.Schema({
    "id": {
       type: String,
       map: "pk"
     }
    "type": {
       type: String,
       map: "sk"
     }
});

const orderSchema = new dynamoose.Schema({
    "userId": {
       type: String,
       map: "pk"
     }
    "orderId": {
       type: String,
       map: "sk"
     }
});
```

**Comment by fishcharlie** [comment]
@alexedwardjones Nope, no plans until you created this issue 😉. **Love** this idea tho. I will prioritize it for v3. Really appreciate the feedback and feature suggestion.

Also, thank you for trying v3. Any feedback (good or bad) is appreciated. Even good feedback tells us that we are almost ready to ship. So really appreciate it. Feel free to continue to create issues or [contact me](https://charlie.fish/contact) with any other feedback.

Thanks again!

**Comment by fishcharlie** [comment]
@alexedwardjones I'm working on planning how this feature is going to work, and have a question regarding the design for it.

Would it be acceptable to have the `map` property represent the name you want to use within your application & Dynamoose, and the actual property name be the name stored in DynamoDB?

For example, your schema would look like the following:

```js
const userSchema = new dynamoose.Schema({
    "pk": {
       type: String,
       map: "id"
     }
    "sk": {
       type: String,
       map: "type"
     }
});
```

Then in DynamoDB the attribute names will be `pk` & `sk`. And within your Items you can store those attributes as `id` & `type`. And when you retrieve items from DynamoDB, they will be converted into `id` & `type` so you can easily access them within your application.

Basically just switching the Schema positions around.

Would this be acceptable to you?

**Comment by alexedwardjones** [comment]
@fishcharlie Yeah, that works. As long as they're accessible based on the interface I supply to the model, then I'm happy either way 😄 

```
interface User extends Item {
  id: string;
  type: string;
}

const UserModel = dynamoose.model<User>(
  process.env.DYNAMODB_TABLE_NAME,
  new dynamoose.Schema({
    "pk": {
       type: String,
       map: "id"
     }
    "sk": {
       type: String,
       map: "type"
     }
});
```

**Comment by fishcharlie** [comment]
@alexedwardjones This work is complete. Is there any chance you could test & review PR #1322 and let me know what you think? Thanks!

**Comment by fishcharlie** [comment]
@alexedwardjones Totally understand if you are busy. But just following up on this. Would love to get this merged in as soon as possible. But really want to ensure you test and review the PR prior to merging to ensure it fits with what you requested.

**Comment by AlisonVilela** [comment]
@fishcharlie  Thanks for implementing this feature.
I wonder if it would be possible to add `prefix` too? and maybe a way to add multiple PK and SK

Something like:
````typescript
const racerSchema = new dynamoose.Schema({
    "pk": {
       type: String,
       map: "id",
       prefix: "racer"
     },
     ...
});

const raceSchema = new dynamoose.Schema({
    "pk": {
       type: String,
       map: "id",
       prefix: "racer"
     },
    "sk": {
       type: String,
       map: "raceId",
       prefix: "race"
     },
     ...
});

const racePartSchema = new dynamoose.Schema({
    "pk": {
       type: String,
       map: "id",
       prefix: "race"
     },
     "sk": {
       type: String,
       map: "raceId",
       prefix: "race"
     },
     "sk_1": {
       type: String,
       map: "partId",
       prefix: "part"
     },
     ...
});
````

To implement a Single Table Design like this AWS Blog post [creating-a-single-table-design-with-amazon-dynamodb](https://aws.amazon.com/blogs/compute/creating-a-single-table-design-with-amazon-dynamodb/)

![image](https://user-images.githubusercontent.com/6244639/174146016-2930b5f9-e403-46d3-94ba-0521ba309dcf.png)

If it's already possible to do something like this sorry for the post

**Comment by fishcharlie** [comment]
@AlisonVilela 

> I wonder if it would be possible to add `prefix` too?

How would this work? Would it just `prefix` all values with a certain prefix string? How would this work on reads from DynamoDB? Or would it only apply to saves? How about a prefix on a key property and you get using that property? Would the value you get on be prefixed with that value before running the query?

Something similar that you might want to look into is the `set` function: https://beta.dynamoosejs.com/guide/Schema#set-function--async-function. This would allow you to set a function that gets run whenever saving an item to DynamoDB, and also when retrieving an item with that property as part of the query. I think that might meet your needs here.

> and maybe a way to add multiple PK and SK

I don't understand what you mean here. In DynamoDB you can only have one primary key, and one sort key. You can't have multiple of both. Just not allowed in DynamoDB.

If you are wanting to combine two attributes that you have into a single `PK` or `SK` attribute, that can be easily achieved using the `combine` type in Dynamoose: https://beta.dynamoosejs.com/guide/Schema#attribute-types. However, it is important to note that the underlying values will also be stored as part of your item.

For example, lets say you had a `raceId` property, and a `partId` property. And you want your `PK` to be the values of both of those properties, separated by a `-`. You could use the Combine type on your `PK` property, and have it combine the raceId and partId values automatically for you. But the underlying item will still include the raceId and partId values as different attributes.

> Using the Schema's Parent would work similar?

Apologies, but I don't understand what you mean here at all.

---

Hopefully this answers your questions.

**Comment by jdrydn** [comment]
@fishcharlie Apologies for reviving a stale issue, is this what you had in mind when suggesting `set` above? (Modified from [your response in a different thread](https://github.com/dynamoose/dynamoose/issues/768#issuecomment-620910174)):

```js
const postSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        "map": "blogId",
        "get": (val) => val.substring("BLOG#".length)
        "set": (val) => `BLOG#${val}`,
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        "map": "postId",
        "get": (val) => val.substring("POST#".length)
        "set": (val) => `POST#${val}`,
    },

    // ... More post attributes here
    // ... Hopefully not 400kb worth of content otherwise
    // ... DynamoDB will throw an Item-too-large err!
});
```

Which would provide the application layer objects like:

```
{
  "blogId": "MY-BLOG-ID",
  "postId": "MY-POST-ID",
  "title": "Hello, world!"
}
```

And write said object to DynamoDB like:

```
{
  "pk": { "S": "BLOG#MY-BLOG-ID" },
  "sk": { "S": "POST#MY-POST-ID" },
  "title": { "S": "Hello, world!" }
}
```

**Comment by fishcharlie** [comment]
@jdrydn Yep. The only thing to consider is if you want to handle use cases where `POST#` doesn't exist and the data got corrupted. Or what happens if the application layer includes `POST#`.

Maybe those are edge cases that you don't care about. But just something to think about.

**Comment by jdrydn** [comment]
@airstrike I recommend you go through [The What, Why, and When of Single-Table Design with DynamoDB](https://www.alexdebrie.com/posts/dynamodb-single-table/) (& [GraphQL, DynamoDB, and Single-table Design](https://www.alexdebrie.com/posts/dynamodb-graphql/) as a good follow-up / implementation / use-case, both) by [Alex DeBrie](https://www.alexdebrie.com/about) to better understand the DynamoDB one-table approach.

Essentially, you can only perform a single CRUD operations against an item in DynamoDB, uniquely defined by either one property (hash key) (`pk`/`id`) or two properties (hash/range keys) (`pk`/`sk`). You can use transactions to perform a single CRUD operation against multiple items in DynamoDB, but that has limitations to consider.

In the example above, I think you'd need multiple Dynamoose schemas to define the Entity, using the `map`/`get`/`set` features per-property to conform to the single DynamoDB table schema you've outlined:

> Entity | HASH (pk) | RANGE (sk)
> -- | -- | --
> User | USER#<USERNAME> | #METADATA#<USERNAME>
> Photo | USER#<USERNAME> | PHOTO#<USERNAME>#<TIMESTAMP>
> Reaction | REACTION#<USERNAME>#<TYPE> | PHOTO#<USERNAME>#<TIMESTAMP>
> Friendship | USER#<USERNAME> | #FRIEND#<FRIEND_USERNAME>

```js
const userSchema = new dynamoose.Schema({
    "pk": {
        "type": String,
        "hashKey": true,
        "map": "username",
        "get": (val) => val.substring("USER#".length)
        "set": (val) => `USER#${val}`,
    },
    "sk": {
        "type": String,
        "rangeKey": true,
        "map": "username",
        // Unsure what would happen here, with multiple properties "mapped" to the same value?
        "get": (val) => val.substring("METADATA#".length)
        "set": (val) => `METADATA#${val}`,
    },
    "email": {
        "type": String,
    },
});
```
